Remove browserify-rails and Teaspoon, use Webpack and Jest outside of Rails

kevinrobinson opened the pull request:
This has two major changes, aimed at improving the developer experience for front end work (from discussion in https://github.com/studentinsights/studentinsights/issues/1161).  First, it removes browserify-rails and builds most JS instead with Webpack outside of Rails altogether.  Second, it removes Teaspoon and adds Jest as the test runner.

For developers, this means running `yarn start` instead of `bundle exec rails s` and then just working like normal.  And running `yarn test` to run tests instead of using Teaspoon.  The compilation time in development is ~500ms (on save), a test run takes ~10 seconds, and a production build takes ~15 seconds.

In this pull request, almost all the complexity is in changes to the tests, and so the core change for development and production mode is best seen in the [first commit](https://github.com/studentinsights/studentinsights/commit/8fddf48b25e61fc14961fcb935bc59dbafd1ae7b).  The third commit cleans out the Travis build a bit too, since we don't need phantom binaries (or particular old versions of bundler).

### What would it take to deploy this?
- [x] get feedback on this approach
- [x] verify this works for other folks (including PDF rendering)
- [x] fix problem with Highcharts
- [x] fix PDF rendering
- [x] check production build bundle sizes
- [x] check react-select and styles
- [x] manually test locally in IE11
- [x] test deploys
- [x] verify gzipping works in production
- [x] verify product works in production in IE
- [x] verify PDF rendering in production
- [x] update README docs


If this seems good, we could ship this next week.

### How would development work?
Running `yarn start` spins up both Rails and Webpack.  Rails runs as normal, but there's a new script tag in `application.hml.erb`, that loads `public/dev/bundle.js`.  Webpack runs in watch mode, building all the JS and putting it there for Rails to read.  There's no live or hot reloading, but JS is recompiled on save and ready on the next page load.

The entrypoint is a file that replicates the ordering that was in `application.js`, and imports all those files as side-effects, where they rebuild the same `window.shared` object.  This has the same effect as just concatenating all the files like Sprockets does (and @alexsoble I should have thought of this earlier!).  We can still keep incrementally migrating to ES6 modules in like we could with browserify-rails for https://github.com/studentinsights/studentinsights/issues/1180.

### How would production deploys work?
There's a separate Webpack config for a production build, and there's a Ruby hook to run that after `rake assets:precompile` and puts it in `public/build`.  The script tag in `application.html.erb` loads that instead of the development bundle in production.  I didn't test it in staging yet, but this should require no changes to deployment.

### How would testing work?
#### Teaspoon
I got a bit stuck here.  I started trying to make Teaspoon work with browserify-rails to start, and can't remember why that didn't work.  Removing it, the tests didn't work anymore since they use import/export.  They were getting compiled to ES5 somehow, but there was no module loader so `require` calls were failing.  This didn't make much sense to me, so I switched gears.

#### Jest
Adding Jest required two main changes related to the test runner.  First, we use the `toContainText` matcher a bit, and that's a `jquery-jasmine` matcher.  The way to add matchers is Jest seems clunky (you can't do it globally, and have to add them inside a `beforeEach` block).  We could do that but here I dropped the matcher and switched these to `expect($(el).text()).toContain('foo')`.  Second, we use instance variables in a few places, most importantly in the `SpecSugar.withTestEl` method.  `this` isn't passed around in Jest, so I changed this to pass a `container` object that the helper code injects the test element into later.  Others I just changed to lexically scoped variables.

After that, there were a bunch of changes needed to the tests, mostly around dependencies that were still in application.js and so not present in the test environment.  There are things like jQuery plugins (autocomplete, datepicker, Highcharts), Sanitize, Cookie, moment and d3.

One thing that's lost in Jest is being able to use Web Inspector to debug tests, but Jest has [support for node-inspector](http://facebook.github.io/jest/docs/en/troubleshooting.html) if we want to add it.

### Other notes
There's a separate `student_report_pdf.js` bundle for `wkhtmltopdf` too.

### Alternatives
#### Could we just make browserify-rails faster?
I looked at browserify-rails and added some logging, but hard a hard time understanding where the slowdown was coming from.  `browserify-rails` is a Sprockets postprocessor, and there were ~2 second delays happening in some other code in between the calls into the Sprockets postprocessor ([logs](https://gist.github.com/kevinrobinson/5dd90c4fde0dd61c969b7a1255b3c4d3)).  Commenting out the body of the class cut the time down to zero, so I don't really understand what was slowing it down.  I read a bit of Sprockets to see if I could understand, but then stopped.

#### Why not use Webpacker?
This would be great!  But it didn't just work last time we tried.  Separately, there's benefits from removing the Ruby/Rails middleman in terms of being able to debug and configure, so I'm probably a bit biased towards that too.  The build steps are just plain JS.

####  What about create-react-app?
There's a bunch of other work to make a single-page app outside of Rails (eg, authentication, routing, styles).  I wanted to avoid that here, while taking a step in that general direction.

jhilde commented:
Whoa @kevinrobinson, the elves were at work last night!

I had been trying a similar switch the rails application.js into an entry point with lots of random imports strategy with webpacker. Was getting some traction but certainly not complete.

Is the structure in jest similar enough to jasmine that fixing some more global dependencies will make all of our tests work? 

Will dig in more at some point. I need to find some of those elves you have though!

kevinrobinson commented:
@jhilde Yeah, tests are all green.  jest is mostly a drop-in replacement, there were two changes related to the test runner (see above).  Most of the work was around moving the dependencies that are still managed in Rails to work in Jest (which doesn't know about them).  That's just an intermediate step to not have to empty out application.js all in one go.

I also noticed this morning that `jasmine.createSpy` still worked, seems that's a niceness for migration ([code](https://github.com/facebook/jest/blob/6f7ece619b2dc722d27038c6077864109fd920f7/flow-typed/npm/jest_v20.x.x.js#L512)) but I converted those fully now too.

alexsoble commented:
Ayeeeeee @kevinrobinson this is great! Congrats + thanks for putting your brainpower into figuring out how to make this migration happen. And in one PR too!

At first I was confused about how we could `import` files that don't `export` anything, like in `legacy.js`, then I did a little StackOverflow digging and found this: 

https://stackoverflow.com/questions/41179828/es6-import-equivalent-of-require-without-exports/41179861#41179861 

... which helped me understand how this works. This strategy of importing all the legacy files as side-effects is awesome and lets us bypass soooo much mechanical work! :D

I was able to verify `yarn start` locally, and could click around and use the web app with the new build process! 🎉 🎉 

One piece that didn't work for me yet was the PDF rendering. Clicking "generate student report" just gave me an infinite spinny circle in my browser tab. You mentioned that there's a separate student_report_pdf.js bundle for wkhtmltopdf too... does that need any explicit loading or starting on the developer's end?

I also tried deploying this branch to the `somerville-teacher-tool-demo` Heroku app, the push failed and here's why:

```
remote:        $ NODE_ENV=production webpack --config ui/config/webpack.prod.js
remote:        sh: 1: webpack: not found
remote:        error Command failed with exit code 127.
remote:        info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
remote:  !
remote:  !     Precompiling assets failed.
remote:  !
remote:  !     Push rejected, failed to compile Ruby app.
remote: 
remote:  !     Push failed
remote: Verifying deploy...
remote: 
remote: !	Push rejected to somerville-teacher-tool-demo.
remote: 
```

Looks like we need to tell Heroku to download webpack before it does anything else to build the rest of the JS? I can dig into this some more tonight or Monday.

Overall, looks like we're within striking distance! @kevinrobinson I like your checklist and I agree that we should be able to ship this next week!

kevinrobinson commented:
@alexsoble Ah thanks!

For the deploy, `webpack` is specified in package.json as a devDependency.  So a first thing to check is whether Heroku runs `yarn install` on a Rails deploy (I think it will).  If not we can run it ourselves [here](https://github.com/studentinsights/studentinsights/pull/1182/files#diff-52cfe83d7f9f1c952a7995dfb0f99341R5).  One wrinkle is that in a production node project you typically wouldn't install `devDependencies`, so we may need to add a flag that tells yarn to install those as well.

For the PDF, yep I think this just isn't working yet and is the last substantive bit.  I thought the other day I checked back on master and had the same problem too, but confirmed now it's this branch.  I don't understand what's up yet, or how to debug.  When we ask `wicked_pdf` to render the PDF, we pass an option telling it to wait for a [magic window.status value](https://github.com/studentinsights/studentinsights/blob/master/app/controllers/students_controller.rb#L71).  That's how it knows to wait until the render is done, so I thought that might be the problem.  But I looked at this just now by adding `&debug` at the end and I can see it's set just fine.  But without `&debug` the `wkhtmltopdf` process hangs and doesn't even ever timeout, I have to kill the process.

kevinrobinson commented:
Added https://github.com/studentinsights/studentinsights/pull/1182/commits/dfd6db7d87b09bc4dc7dbea231bb28f5ee44d553, which removes all of application.js except for jquery, jquery_ujs (not sure if we use that), and jquery-ui.

kevinrobinson commented:
Okay, https://github.com/studentinsights/studentinsights/pull/1182/commits/3779ff221c440b55d8370fe44c6f5a0a93d45a14 addresses the PDF issue.  It was from two things: the charset wasn't set and so UTF characters like π would throw, and I removed the `Function.bind` polyfill since I assumed any shims would be handled by the `react-app` Babel preset.  This doesn't go that far back though, so I added back `es5-shim` and imported that only in the PDF file.

FWIW, debugging this worked best by grabbing the `whhtmltopdf` command on the Rails output, running that directly on the command line, and then manually updating the HTML file with new JS while iterating.  There's a `javascript-delay` option which delays before exiting, and a `debug-javascript` option that lets your `console.log` output show up on the command line when running `wkhtmltodpf` directly.

kevinrobinson commented:
### Development
Aside from compilation times, this also speeds up the local development experience quite a bit because of reducing the number of files the browser has to fetch (it only makes 6 requests in parallel and so has to block even requesting them all).  Check out the bottom line here: the number of files, total MB and time.  This is a hard reload of a page that's already been compiled and nothing has changed.

Before:
<img width="693" alt="screen shot 2017-10-24 at 7 50 22 am" src="https://user-images.githubusercontent.com/1056957/31941712-c28d6702-b890-11e7-81bc-5632aa0717f4.png">

After:
<img width="699" alt="screen shot 2017-10-24 at 7 49 41 am" src="https://user-images.githubusercontent.com/1056957/31941710-c0c77566-b890-11e7-9e49-a7ca0ac0b4dd.png">

### Production
This also improves the production build's bundle size.  Before, it was:
- application.js 2.7MB (909kb gzipped)

After it is:
- bundle.js 974kb (310kb gzipped)
- application.js 795kb (206k gzipped)

There's plenty of ways to optimize that more (right now we include all of jQuery UI in application.js when we only need the datepicker and autocomplete).  But I think this is an improvement in both dev and production and good to ship in this respect.

alexsoble commented:
👏 👏 👏  Fantastic improvements right off the bat @kevinrobinson!

I started working on deploying to Heroku (demo site) on this branch: https://github.com/studentinsights/studentinsights/tree/deploy-webpack-jest-to-heroku

If it makes sense to you, once I get the branch to deploy to the demo site I can open a PR from my branch into this one. Or happy to go with a different deploy strategy / merge strategy if you prefer!

alexsoble commented:
Deployed to https://somerville-teacher-tool-demo.herokuapp.com/! Deploy is live, cool to see `<script src="/build/bundle.c8c8326f17126602d8b8.js"></script>` at the bottom of the page. 

And maybe it's just placebo but the app definitely feels ... faster.

Opened #1187 to merge the deployment branch into this PR.

kevinrobinson commented:
@alexsoble Awesome, thanks!  Do you think this is good enough to check off "verify deploys" in the issue's checkbox then?

kevinrobinson commented:
Also your demo warning banner is amazing :)

alexsoble commented:
@kevinrobinson Thank you! Tried to keep it nice and subtle!

alexsoble commented:
@kevinrobinson I'll leave it up to you to decide if we want to check off "verify deploys". Clearly we can get it deployed by moving dependencies out of devDependencies but I'm happy to work towards a more elegant solution so that we fix it in a 👌 way.

kevinrobinson commented:
Added `Object.assign` polyfill with https://github.com/studentinsights/studentinsights/pull/1182/commits/b7b716b9de2ceed9fde16d1bb1f81cebd305a1ee to address bug affecting IE11.  This is tricksy since the call to Object.assign isn't in our code but is produced in the Babel transpilation process, and so our eslint rule from https://github.com/studentinsights/studentinsights/pull/1002 can't detect it.  This also removes the need for the rule in https://github.com/studentinsights/studentinsights/pull/1002 altogether.

kevinrobinson commented:
Manual testing in IE surfaced a bunch of small UI issues (https://github.com/studentinsights/studentinsights/issues/1189 and https://github.com/studentinsights/studentinsights/issues/1190 https://github.com/studentinsights/studentinsights/issues/1191), but these are unrelated and not going to address them here.

It also revealed that the STAR charts have been broken for a week and a half, but that's unrelated and opened https://github.com/studentinsights/studentinsights/pull/1158 for that, not going to address it here either.

kevinrobinson commented:
Verified demo deploys work with the change in https://github.com/studentinsights/studentinsights/pull/1187#issuecomment-339048677, added in https://github.com/studentinsights/studentinsights/pull/1182/commits/bc87b46620d174fb9a47ffb5d0307204ed182334.

kevinrobinson commented:
Okay, verified that PDFs are good in the production demo site.  There are some minor style differences, but not important enough to look further right now.  Note there's one extra label and the legend shape is different.  Maybe I grabbed the wrong minor version (we're two major versions behind).

### Before
<img width="733" alt="screen shot 2017-10-25 at 9 01 11 am" src="https://user-images.githubusercontent.com/1056957/32000031-4e8e0cfc-b963-11e7-8b34-6b59515e3faa.png">

### After
<img width="736" alt="screen shot 2017-10-25 at 9 01 06 am" src="https://user-images.githubusercontent.com/1056957/32000032-5073d29a-b963-11e7-9250-7be8d459d74f.png">

kevinrobinson commented:
@alexsoble @jhilde This should be ready to go now, just needs documentation updates.  The gist is `yarn start` in one terminal to start Rails and Webpack and `yarn test` in another to run the tests.  And `rm -rf node_modules && yarn install` to setup the first time (or on dependency changes).

Can you both check this out at some point (the first commit https://github.com/studentinsights/studentinsights/commit/8fddf48b25e61fc14961fcb935bc59dbafd1ae7b is the important one) and share feedback or give a thumbs up when you feel good about this?

After that I can ping everyone with a heads up about this, and can help resolve merge conflicts on any open PRs (but I don't think we will have many).

alexsoble commented on lib/tasks/assets.rake:
```diff
@@ -2,5 +2,7 @@
-  exec 'yarn run build'
+  # Run yarn install first to force devDependencies to be installed, 
+  # so that we can run the build process
+  exec 'YARN_PRODUCTION=false yarn install --frozen-lockfile && yarn run build'
```
👍

alexsoble commented:
I feel good about this. Pulled the latest changes, ran `yarn start` again, and explored the build. I also wrapped my mind around the legacy.js `import` strategy. 👍 

Open PRs that may have merge conflicts—especially in the JS specs, since many of the-features side `.jsx` files are amazingly untouched by this PR—are #1178, #1165, and #1154.

kevinrobinson commented:
Docs are updated in https://github.com/studentinsights/studentinsights/pull/1182/commits/e533d7494c258cc554ce663504c80d503b8b5bbf.  I made progress on Docker in https://github.com/studentinsights/studentinsights/pull/1182/commits/d972ca47087a7ea597e2ad08c21bedca2340fcc9, but this branch can't yet build the container, `yarn install` fails saying "Yarn requires Node.js 4.0 or higher to be installed."

kevinrobinson commented:
The Docker setup is working in https://github.com/studentinsights/studentinsights/pull/1182/commits/60ddf27367bb9bff98efe82980067f8729d7f6f7, but performance is not good on my computer.  I'm not sure if this is a regression or just related to the large number of files.  For me, `docker-compose build` took 10m.  The initial Webpack build running `yarn start` inside the container took 71246ms.

I added a `:cached` hint which helped a little, bringing a page load from ~12 seconds to ~9 seconds (with no changes or anything, that's just time to load the page).

Going to push this up, consider this done, and ask about removing Docker support altogether in https://github.com/studentinsights/studentinsights/issues/1209.

kevinrobinson commented:
🚢ing cc @alexsoble
